SU2_CFD/src/solvers/CSolver.cpp

```diff
@@ -2065,78 +2066,79 @@ void CSolver::AdaptCFLNumber(CGeometry **geometry,
-      New_Func += solverFlow->GetRes_RMS(iVar);
-    }
-    if ((iMesh == MESH_0) && (config->GetKind_Turb_Model() != NONE)) {
-      for (unsigned short iVar = 0; iVar < solverTurb->GetnVar(); iVar++) {
-        New_Func += solverTurb->GetRes_RMS(iVar);
+      New_Func = 0.0;
+      for (unsigned short iVar = 0; iVar < solverFlow->GetnVar(); iVar++) {
+        New_Func += solverFlow->GetRes_RMS(iVar);
+      }
+      if ((iMesh == MESH_0) && solverTurb) {
+        for (unsigned short iVar = 0; iVar < solverTurb->GetnVar(); iVar++) {
+          New_Func += solverTurb->GetRes_RMS(iVar);
+        }
       }
-    }
 
-    /* Compute the difference in the nonlinear residuals between the
-     current and previous iterations. */
+      /* Compute the difference in the nonlinear residuals between the
+       current and previous iterations. */
 
-    NonLinRes_Func = (New_Func - Old_Func);
-    NonLinRes_Series[NonLinRes_Counter] = NonLinRes_Func;
+      NonLinRes_Series[NonLinRes_Counter] = New_Func - Old_Func;
 
-    /* Increment the counter, if we hit the max size, then start over. */
+      /* Increment the counter, if we hit the max size, then start over. */
 
-    NonLinRes_Counter++;
-    if (NonLinRes_Counter == Res_Count) NonLinRes_Counter = 0;
+      NonLinRes_Counter++;
+      if (NonLinRes_Counter == Res_Count) NonLinRes_Counter = 0;
 
-    /* Sum the total change in nonlinear residuals over the previous
-     set of all stored iterations. */
+      /* Sum the total change in nonlinear residuals over the previous
+       set of all stored iterations. */
 
-    NonLinRes_Value = New_Func;
-    if (config->GetTimeIter() >= Res_Count) {
-      NonLinRes_Value = 0.0;
-      for (unsigned short iCounter = 0; iCounter < Res_Count; iCounter++)
-        NonLinRes_Value += NonLinRes_Series[iCounter];
-    }
+      NonLinRes_Value = New_Func;
+      if (config->GetInnerIter() >= Res_Count) {
+        NonLinRes_Value = 0.0;
+        for (unsigned short iCounter = 0; iCounter < Res_Count; iCounter++)
+          NonLinRes_Value += NonLinRes_Series[iCounter];
+      }
 
-    /* If the sum is larger than a small fraction of the current nonlinear
-     residual, then we are not decreasing the nonlinear residual at a high
-     rate. In this situation, we force a reduction of the CFL in all cells.
-     Reset the array so that we delay the next decrease for some iterations. */
+      /* If the sum is smaller than a small fraction of the current nonlinear
+       residual, then we are not decreasing the nonlinear residual at a high
+       rate. In this situation, we force a reduction of the CFL in all cells.
+       Reset the array so that we delay the next decrease for some iterations. */
 
-    if (fabs(NonLinRes_Value) < 0.1*New_Func) {
-      NonLinRes_Counter = 0;
-      for (unsigned short iCounter = 0; iCounter < Res_Count; iCounter++)
-        NonLinRes_Series[iCounter] = New_Func;
+      if (fabs(NonLinRes_Value) < 0.1*New_Func) {
```

Comment:
This might be too stringent of a condition and in most of my runs it would reset the CFL every 100 iterations, even if the solution was converging. Since the chosen numbers of 100 iterations and 0.1 reduction are based on experience, perhaps relaxing this might be a good idea. I'll try a couple of different values, but I wanted to check to see if this is something others have come across as well

These plots are from running with `CFL_ADAPT_PARAM= ( 0.1, 1.1, 1.0, 100.0 )` on a naca0012 with an adapted grid. Similar trends are seen in the standard (NASA TMR) grids as well. 

![Avg CFL_convergence](https://user-images.githubusercontent.com/30271435/103041628-ca027100-452b-11eb-99ee-ebdcfbb35d87.png)
![relrms Rho _convergence](https://user-images.githubusercontent.com/30271435/103041629-ca9b0780-452b-11eb-8443-9c6f0e31e865.png)

Reply:
Yeah I think this would be a problem especially for very large cases.
I thought of using a smoothness criteria that would be "self normalizing" for example counting the number of sign inversions in the history to detect flip-flopping.

Reply:
Yeah, I played around with a couple of different values and there isn't really a universal value that works. A value of 0.01 for the reduction prevented some of the CFL  resetting, but larger cases still have the same issue. I don't think it makes a huge difference to the overall time to solution though. 

A smoothness criteria would be better but I am also okay with moving forward with this.

Reply:
Before Brian's fix this non linear criteria did not affect steady state problems, now it can be a bit annoying depending on how aggressive is the increase factor... I'll try to write something simple based on sign inversions.